### DISCUSSION TITLE
Is there a way to map a house or a room without a LiDAR?

### DISCUSSION THREAD

**Discussion by Illyvia** [iRobotEducation/create3_docs#434] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Old computer running ubuntu

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.2

### Which RMW is your robot running?

FastRTPS

### Is the robot connected to a network? If so, what is the network type?

home network

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

Bluetooth (default)

### Describe your question.

Is there a way to map a room or a house with the create 3 without a LiDAR? (Such as the ir sensors) If so, can the create 3 go to parts of the map controlled by python? Not sure this was possible, just wanted to ask.

**Comment by alsora** [comment]
The Create 3 is a mobile base and it can be used to develop and test robotics algorithms on top of it.

It provides some built-in functionalities and "drive to a pose" is one of them https://iroboteducation.github.io/create3_docs/api/drive-goals/.
These navigation action use an inertial localization system based on fusing IMU, wheel encoders and optical sensor, described here https://iroboteducation.github.io/create3_docs/api/odometry/.

If you want to map an environment, you will need a SLAM algorithm.
There's plenty of SLAM algorithms available open-source, but they all rely on additional sensors such as a LIDAR or a camera.
Developing your own SLAM algorithm is not an easy task and doing so with only the sensors available on the Create 3 is going to be even harder, but it's doable and, as a motivating point, you can consider that the iRobot Roomba I3 (which uses the same hardware as the Create 3) is indeed capable of doing it https://www.theverge.com/2022/3/10/22967391/irobot-roomba-i3-update-smart-mapping.

**Reply by Illyvia** [reply]
Thanks for your quick reply. Could you list a SLAM algorithm that could maybe work with a usb webcam? And if not, how hard would it be to develop a SLAM algorithm?

**Reply by alsora** [reply]
To run with a USB webcam you will need a monocular SLAM algorithm; most algorithms using cameras usually have stereo pairs.
Developing a SLAM algorithm from scratch requires non-trivial competencies in computer vision, robotics and optimziation theory.

I think that you will be better off with an existing algorithm rather than writing your own, at least to understand all the pieces involved.
This paper contains a relatively recent overview of open-source visual SLAM algorithms: https://arxiv.org/abs/2107.07589

We provide a demo and instructions for running SLAM with Create 3 and a LIDAR in case you are interested: https://github.com/iRobotEducation/create3_examples/tree/galactic/create3_lidar

**Reply by Illyvia** [reply]
Thanks very much for the info!

**Reply by Illyvia** [reply]
I have one more question: Does the Navigate To Position command use the ir sensors to navigate, or just assume that there are no obstacles? Thanks

**Reply by alsora** [reply]
@TheInventor2023 all the drive actions mentioned in the link above (navigate to position included) assume that there are no obstacles.

If you want to handle obstacles in a smart way you will need to write a more complex behavior yourself.

**Reply by Illyvia** [reply]
Thanks, and when using LiDAR, do you need ubuntu on both the raspberry pi and the computer?

**Reply by shamlian** [reply]
From my personal experience, the ROS 2 experience is most seamless when running Ubuntu. If you're plugging sensors and the robot into the Raspberry Pi, you'll likely want to install Ubuntu on it (unless you want to be building things from source). Similarly, if you're planning to control the robot via ROS 2 on your external computer, Ubuntu (either native or a container) is likely the way to go. If you're just visualizing on your external computer, you could get away with not installing Ubuntu (for example, a visualizer like Foxglove Studio can run in a browser).
